dynamic_instrumentation/src/main.rs

```diff
@@ -0,0 +1,413 @@
+    }?;
+    for profile_dir in target_read_dir {
+        let profile_dir = profile_dir?;
+        if !profile_dir.file_type()?.is_dir() {
+            continue;
+        }
+        for artifact in profile_dir.path().join("deps").read_dir()? {
+            let artifact = artifact?;
+            if !artifact.file_type()?.is_file() {
+                continue;
+            }
+            let file_name = artifact.file_name();
+            let file_name = file_name.to_str();
+            for crate_target in &info.crate_targets {
+                let prefix = format!("{}-", crate_target.crate_name.as_str());
+                // [`Path::starts_with`] only checks whole components at once,
+                // and `OsStr::starts_with` doesn't exist yet.
+                if file_name
+                    .map(|name| name.starts_with(&prefix))
+                    .unwrap_or_default()
+                {
+                    let artifact = artifact.path();
+                    if artifact.is_executable() {
+                        fs_err::remove_file(&artifact)?;
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn main() -> eyre::Result<()> {
+    color_eyre::install()?;
+    env_logger::init();
+
+    let rustc_wrapper_var = "RUSTC_WRAPPER";
+    let instrument_info_var = "C2RUST_INSTRUMENT_INFO";
+
+    let own_exe = env::current_exe()?;
+
+    let wrapping_rustc = env::var_os(rustc_wrapper_var).as_deref() == Some(own_exe.as_os_str());
+    if wrapping_rustc {
+        let sysroot = get_sysroot()?;
+        let at_args = env::args().skip(1).collect::<Vec<_>>();
+        let crate_target = CrateTarget::from_rustc_args(&at_args)?;
+        let info = env::var(instrument_info_var)?;
+        let info = serde_json::from_str::<InstrumentInfo>(&info)?;
+        let should_instrument = info.crate_targets.contains(&crate_target);
+        if should_instrument {
+            instrument_rustc(at_args, &sysroot, &info.metadata)?;
+        } else {
+            let status = passthrough_rustc(&at_args, &sysroot)?;
+            exit_with_status(status);
+        }
+    } else {
+        let Args {
+            metadata,
+            cargo_args,
+        } = Args::parse();
+
+        let cargo_metadata = MetadataCommand::new().exec()?;
+        let crate_targets = CrateTarget::from_metadata(&cargo_metadata)?;
+        let info = InstrumentInfo {
+            crate_targets,
+            metadata,
+        };
+
+        delete_metadata_and_dependencies(&info, &cargo_metadata)?;
+
+        // We could binary encode this, but it's likely very short,
+        // so just json encode it, so it's also human readable and inspectable.
+        let info = serde_json::to_string(&info)?;
+
+        let cargo = env::var_os("CARGO").unwrap_or_else(|| "cargo".into());
+        let status = Command::new(cargo)
+            .args(cargo_args)
+            .env(rustc_wrapper_var, &own_exe)
+            .env(instrument_info_var, info)
```

Comment:
Environment variables have max lengths, so I'm concerned about this string being too long to fit here. Ideally we can avoid passing this information entirely.

Reply:
I think in practice it should be fine, because we only pass the metadata path and the targets of the main package (i.e. the `lib`, `bin`, etc. crates), which there usually aren't very many of.  But if there's a way we can avoid this, that'd be better, though doing the `cargo metadata` lookup in each `rustc` invocation is not the best option, as that repeats a lot of work.

Reply:
Not an issue anymore, as now we only pass the metadata path, which shouldn't (can't?) be too long.